fitbenchmarking/parsing/mantiddev_parser.py

```diff
@@ -0,0 +1,236 @@
+from mantid.api import FunctionDomain1DVector as FDV
+from fitbenchmarking.parsing.fitbenchmark_parser import FitbenchmarkParser
+
+
+class MantidDevParser(FitbenchmarkParser):
+    """
+    Parser for a Mantid problem definition file.
+    """
+
+    def __init__(self, filename, options):
+        self._params_dict = None
+        self._mantid_function = None
+        self._jac = None
+        self._N_x = 0
+        self._cache_x = None
+        super().__init__(filename, options)
+
+    def _set_additional_info(self) -> None:
+        """
+        Sets any additional info for a fitting problem.
+        """
+        if self.fitting_problem.multifit:
+            self.fitting_problem.additional_info['mantid_ties'] \
+                = self._parse_ties()
+
+    def parse(self):
+
+        fp = super().parse()
+
+        self._set_jacobian(fp)
+        return fp
+
+    def _set_jacobian(self, fp) -> None:
+        """
+        Sometimes mantid will give the error
+        RuntimeError: Integration is not implemented for this function.
+        this try except tests if the error occurs and then only
+        assigns the jacobian if it passes.
+        :param fp: fitting problem
+        """
+        if self._is_multifit():
+            # currently cannot do Jacobian and multifit
+            return
+        # need to trim x data to the correct range for Jacobian
+        i0 = 0
+        iN = len(fp.data_x)
+        if fp.start_x:
+            i0 = np.argmax(fp.data_x >= fp.start_x)
+        if fp.end_x:
+            # returns a list of lists if more than one match,
+            # otherwise an int is returned
+            iN = np.where(fp.data_x <= fp.end_x)
+            if not isinstance(iN, int):
+                iN = iN[0][-1]
+
+        x_data = fp.data_x
+        x_data = x_data[i0:iN + 1]
+
+        # cache the x values for later
+        self._cache_x = FDV(x_data)
+        self._N_x = len(x_data)
+        self._jac = np.zeros((self._N_x, len(self._params_dict.keys())))
+
+        try:
+            _ = self._jacobian(x_data, self._params_dict.values())
+            fp.jacobian = self._jacobian
+        except RuntimeError:
+
+            return
+
+    def _jacobian(self, _x, *args):
+        # pylint: disable=unused-argument
+        """
+        Extracts the Jacobian from Mantid
+        WARNING: Gaussians are known to be incorrect
+
+        :param _x: the x values for the problem (assume they have not changed)
+        :param args: the input parameters for the fit
+        :return a matrix of the Jacobian
+        """
+        for param, key in zip(*args, self._params_dict.keys()):
+            self._mantid_function[key] = param
+        # get mantid Jacobian
+        J = self._mantid_function.functionDeriv(self._cache_x)
+        # set np Jacobian values
+        for i in range(self._N_x):
+            for j in range(len(self._params_dict.keys())):
+                self._jac[i, j] = J.get(i, j)
+        return self._jac
```

Comment:
I'm a bit concerned about using the same object here. If we call this twice on 2 different inputs both will be the jacobian of the second input?

```
x1 = ...
j1 = fitting_prob.jacobian(x1, params)
x2 = ...
j2 = fitting_prob.jacobian(x2, params)

j1 == j2
```

Will this be an issue on finite difference hessians?

Reply:
Is there a way to test finite difference Hessians? I can just try it and see if everything goes crazy. 

I have done it this way so we dont need to create a new Jacobian every time. I could change that or use a deepcopy

Reply:
If you run with `software=scipy` or `ralfit` and set `hes_method=numdifftools` or `scipy`. It makes sense to save the cost of allocating a new array but it might not be worth it here as they'd have to be copied later anyway.

Reply:
In the table the one column with the awful times is using the fd hessians. I suspect, it's finding that H=0 so just doing something like gradient descent which we'd expect to perform poorly.

It's probably a hole in the docs that we don't say which solvers are able to use hessians and which aren't. We also need to have a better description of the column titles (a `j:...` indicates the jacobian used, if it's omitted, no jacobian was given to the solver. Similarly for hessians with `h:`)

For scipy we have:
```
    hessian_enabled_solvers = ['Newton-CG',
                               'trust-ncg',
                               'trust-exact',
                               'trust-krylov',
                               'trust-constr',
                               'dogleg']
```

Reply:
I just changed the code to
```
        _jac = np.zeros((self._N_x, len(self._params_dict.keys())))
        for i in range(self._N_x):
            for j in range(len(self._params_dict.keys())):
                _jac[i, j] = J.get(i, j)
        return _jac
```
and the run time is slower! I got 95.77 and 160.1 for mantid and mantidev respectively, for the Newton-CG: J:an analytic h:numdifftools central

Reply:
The slow down is dues to generating the hessian from scratch every time. However, even the NIST problems are very slow. This combined with the above suggests that the mantid jacobians are being copied (the slowness is similar to  NIST's)

Reply:
Yeah it must be being copied somewhere! I think I'll add a comment so if a new hessian is added and doesn't copy, it's a bit easier to track down.

I'll put the new parser on the regression tests too.